credentials/sts/sts.go

```diff
@@ -0,0 +1,427 @@
+		return nil, err
+	}
+
+	// Load the system roots to validate the certificate presented by the STS
+	// endpoint during the TLS handshake.
+	roots, err := loadSystemCertPool()
+	if err != nil {
+		return nil, err
+	}
+
+	return &callCreds{
+		opts:   opts,
+		client: makeHTTPDoer(roots),
+		bs:     makeBackoffStrategy(),
+	}, nil
+}
+
+// defaultBackoffStrategy returns an exponential backoff strategy based on the
+// default exponential backoff config, but with a smaller BaseDelay (because the
+// former is meant more for connection establishment).
+func defaultBackoffStrategy() backoff.Strategy {
+	config := grpcbackoff.DefaultConfig
+	config.BaseDelay = 100 * time.Millisecond
+	return backoff.Exponential{Config: config}
+}
+
+// httpDoer wraps the single method on the http.Client type that we use. This
+// helps with overriding in unittests.
+type httpDoer interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+func makeHTTPClient(roots *x509.CertPool) httpDoer {
+	return &http.Client{
+		Timeout: stsRequestTimeout,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs: roots,
+			},
+		},
+	}
+}
+
+// validateOptions performs the following validation checks on opts:
+// - tokenExchangeServiceURI is not empty
+// - tokenExchangeServiceURI is a valid URI with a http(s) scheme
+// - subjectTokenPath and subjectTokenType are not empty.
+func validateOptions(opts Options) error {
+	if opts.TokenExchangeServiceURI == "" {
+		return errors.New("empty token_exchange_service_uri in options")
+	}
+	u, err := url.Parse(opts.TokenExchangeServiceURI)
+	if err != nil {
+		return err
+	}
+	if u.Scheme != "http" && u.Scheme != "https" {
+		return fmt.Errorf("scheme is not supported: %s. Only http(s) is supported", u.Scheme)
+	}
+
+	if opts.SubjectTokenPath == "" {
+		return errors.New("required field SubjectTokenPath is not specified")
+	}
+	if opts.SubjectTokenType == "" {
+		return errors.New("required field SubjectTokenType is not specified")
+	}
+	return nil
+}
+
+// callCreds provides the implementation of call credentials based on an STS
+// token exchange.
+type callCreds struct {
+	opts   Options
+	client httpDoer
+	bs     backoff.Strategy
+
+	// Cached accessToken to avoid an STS token exchange for every call to
+	// GetRequestMetadata.
+	mu          sync.Mutex
+	cachedToken *tokenInfo
+}
+
+// GetRequestMetadata returns the cached accessToken, if available and valid, or
+// fetches a new one by performing an STS token exchange.
+func (c *callCreds) GetRequestMetadata(ctx context.Context, _ ...string) (map[string]string, error) {
+	if md := c.metadataFromCachedToken(); md != nil {
+		return md, nil
+	}
+	req, err := c.constructRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	// Send the request with exponential backoff and retry. Even though not
+	// retrying here is OK, as the connection attempt will be retried by the
+	// subConn, it is not very hard to perform some basic retries here.
```

Comment:
It's not hard, but it may go against some principles of gRPC.  Consider the first connection attempt in particular -- if we get an error here caused by misconfiguration, we could return that error immediately and RPCs would fail right away at startup.  But if we retry and wait for 20 seconds or a minute to elapse before failing, then even fail-fast RPCs will stall for that duration, when they have no hope of succeeding.  Also, the error returned is (currently) the context error and not the last encountered error, which hurts debuggability (but this can be fixed).

Yes we will churn the server connection if a transient error occurs here, but that seems like a tradeoff worth making to me.

Reply:
I think that would be best, yes.

Reply:
Done.